app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistManager.java

```diff
@@ -82,6 +86,51 @@ public Completable updateJoin(final long playlistId, final List<Long> streamIds)
         })).subscribeOn(Schedulers.io());
     }
 
+    public Maybe<List<PlaylistStreamEntry>> clearWatched(final long playlistId,
+                                                         final boolean removePartiallyWatched) {
+        return Flowable.combineLatest(getPlaylistStreams(playlistId),
+                    streamHistoryDAO.getHistoryStreamIdsSortedById(),
+                    (streams, historyIds) -> {
+                        final List<PlaylistStreamEntry> notWatched;
+                        if (removePartiallyWatched) {
+                            notWatched = streams.stream()
+                                    .filter(s -> !historyIds.contains(s.getStreamId()))
```

Comment:
Even though history ids are sorted, this won't do a binary search to determine whether `s.getStreamId()` is contained, since the type of `historyIds` is just `List`. This can slow down the algorithm from `O(n*log(n))` to `O(n^2)`, which is unacceptable for big playlists/histories. So use `Collections.binarySearch()` instead

Reply:
Like this?
```java
final List<PlaylistStreamEntry> notWatched;
if (removePartiallyWatched) {
   notWatched = streams.stream()
      .filter(s -> Collections.binarySearch(historyIds, s.getStreamId()) < 0)
      .collect(Collectors.toList());
} else {
   notWatched = streams.stream()
      .filter(s -> Collections.binarySearch(historyIds,  s.getStreamId()) < 0 || s.getProgressTime() > 0)
      .collect(Collectors.toList());
}
```

Reply:
Yes, looking good